### ISSUE TITLE
Too Easy to Win, Too Hard to Lose

### ISSUE THREAD

**Issue by dbear496** [lincity-ng/lincity-ng#220]
There are several subtle parts of the game that can be exploited to make getting a sustainable economy very easy:

- The player's money balance is floored at -2G. It turns out it is possible to have a sustainable economy with negative cash flow by exploiting this.
- Markets burn waste, and this can be used instead of dumping on tips or recycling.
- It is possible to "deal with" pollution and/or starvation just by building high-birthrate residences to keep the population up.
- Homeless population does not eat, cannot die, and does not count as unemployment.
- Mills are kinda OP, so industry isn't really necessary.
- And the only way to lose is by every single person starving, which is easily preventable by building a few farms and wells -- or even just one farm and one well.

I realized this while I was fixing #182, which entailed testing the bulldoze function. So I was randomly scribbling the bulldoze tool across the Rocket_98 map, and I noticed that even after doing so, the economy was nearly sustainable -- only there were a few buildings without fire cover.

So I started a new game to see what kind of degenerate city could still be sustainable, and...it's pretty bad. Turns out the faster I can run out of money, the faster I can be sustainable. Getting 50 tech is the hardest part, and that just takes a few schools, some farms, and some mills. This city has no electricity. All the waste is burned by markets in the north of the map. People are starving, but that comes to equilibrium with the birth rate. The whole city has railroads, but the 566k bill isn't a problem because of the money floor. In fact, the net yearly income is -1.5M, but as I said, not a problem. I just need enough goods to run the schools and firestations, and that is supplied by 30 mills and 114 farms.

**Comment by v1993** [comment]
I suppose an easy start would be to require positive money balance, as flooring is definitely syupposed to be a safeguard of sorts and not an (ab)usable mechanic anyways. It makes no sense to have a ""sustainable economy"" that doesn't use up momey because of a technicality.

Other than that, IMO requirements should be imposed on having less than a certain % of population be homeless and some similar limit on unnatural deaths - that is, too many homeless or too many unnatural deaths should disrupt the sustainable economy progress. Surprised this isn't in place already.

As for losing being too hard - I don't know, I don't really have an issue with that; being in deep negative balance with non-functional infrastructure is already sort of a soft loss condition (balance flooring aside), even if game does not explicitly announce it.

**Comment by dbear496** [comment]
Thanks for the input.

To fix the money floor exploit, I can think of four options, which are not mutually exclusive:

1. Disable sustainability if the money is at the floor.
2. Disable sustainability if money is negative.
3. Lose the game if the money floor is hit
4. Disallow spending if the money floor is hit. This means construction and bulldozing is disabled, and universities, schools, fire stations, hospitals, sports, transport, and wind turbines will be turned off if there are insufficient funds. This would likely make sustainability impossible if the money floor is hit. I'm not sure there is a way to disable spending for unnatural deaths, unemployment, or interest.

I really like number 4, but it would be the most work to implement. Number 4 could also mitigate overflow exploits where the player spends so much in a single year that it overflows a 32-bit integer and makes the money very positive.

I agree with Valeri about the soft-loss condition. I guess the only reason to have a hard loss condition is to communicate to the player that it is near-impossible instead of letting them hopelessly try to resurrect the economy.

For homelessness, perhaps a good requirement for sustainability should be there are no shanties -- and then disable living in communes. For this to work, I would have to fix the issue with the last shanty not disappearing. Also there should be a good way to evacuate/bulldoze residences without initiating a shanty explosion or mass starvation.

For unemployment, perhaps this should cause homelessness (because unemployed people can't pay rent).

**Comment by v1993** [comment]
I'm personally in favor of disabling sustainability at negative money; integer underflow could be mitigated separately with checking for it if it is actually an issue. I do like suggestion 4, and would even suggest going a step further and require a source of immediate income to keep things running if you have negative money, but it definitely would require significant work. So, a mix of 2+3+4 would be ideal IMO and can also be done one step at a time.

Regarding shanties - sounds good to me, although I suppose a sufficiently dedicated player could fill the map with random buildings to prevent these from being built (the basic road in a pattern that leaves no free 2x2 spaces probably would be the cheapest version).

**Comment by dbear496** [comment]
I just looked into it,and I don't think there's a good way to exploit integer underflow for money. Someone would have to spend almost 150M instantly. The most feasible way I think there is to do that is to build 625 rockets so the upkeep cost is 150M at the end of the year. But rockets can't be built on credit. Other things can become very expensive but only if the tech level is very high on the order of ten-thousands.

**Comment by dbear496** [comment]
That's a good point about blocking shanties. There will have to be a check for that. I remember before I understood how to play that I made that road pattern leaving no 2x2 spaces to avoid shanties from springing up. It was very annoying because I had to pause the game in order to build anything.

**Comment by farooqkz** [comment]
What is status of this, or the development overall?

**Comment by dbear496** [comment]
> What is status of this, or the development overall?

kind of a vague question tbh. Obviously this specific issue isn't fixed -- at least not fully; iirc I only fixed the negative money part so far. As far as developement overall, I haven't had a chance to work on this project for the past ~3 months. I got a bit stalled out trying to do some frontend-backend separation.

**Comment by farooqkz** [comment]
> > What is status of this, or the development overall?
> 
> kind of a vague question tbh. Obviously this specific issue isn't fixed -- at least not fully; iirc I only fixed the negative money part so far. As far as developement overall, I haven't had a chance to work on this project for the past ~3 months. I got a bit stalled out trying to do some frontend-backend separation.

Yeah I wanted to know these and if there is a plan. Now I know. Thanks for developing this game :)
